lib/bolt/inventory/inventory2.rb

```diff
@@ -0,0 +1,264 @@
+            ts = resolve_name(name)
+            ts.map do |t|
+              target = create_target(t)
+              target.inventory = self
+              target
+            end
+          end
+        end
+      end
+      private :expand_targets
+
+      def set_vars_from_hash(tname, data)
+        if data
+          # Instantiate empty vars hash in case no vars are defined
+          @target_vars[tname] ||= {}
+          # Assign target new merged vars hash
+          # This is essentially a copy-on-write to maintain the immutability of @target_vars
+          @target_vars[tname] = @target_vars[tname].merge(data).freeze
+        end
+      end
+      private :set_vars_from_hash
+
+      def set_facts(tname, hash)
+        if hash
+          @target_facts[tname] ||= {}
+          @target_facts[tname] = Bolt::Util.deep_merge(@target_facts[tname], hash).freeze
+        end
+      end
+      private :set_facts
+
+      def add_node(current_group, target, desired_group, track = { 'all' => nil })
+        if current_group.name == desired_group
+          # Group to add to is found
+          t_name = target.name
+          # Add target to nodes hash
+          current_group.nodes[t_name] = { 'name' => t_name }.merge(target.options)
+          # Inherit facts, vars, and features from hierarchy
+          current_group_data = { facts: current_group.facts,
+                                 vars: current_group.vars,
+                                 features: current_group.features }
+          data = inherit_data(track, current_group.name, current_group_data)
+          set_facts(t_name, @target_facts[t_name] ? data[:facts].merge(@target_facts[t_name]) : data[:facts])
+          set_vars_from_hash(t_name, @target_vars[t_name] ? data[:vars].merge(@target_vars[t_name]) : data[:vars])
+          data[:features].each do |feature|
+            set_feature(target, feature)
+          end
+          return true
+        end
+        # Recurse on children Groups if not desired_group
+        current_group.groups.each do |child_group|
+          track[child_group.name] = current_group
+          add_node(child_group, target, desired_group, track)
+        end
+      end
+      private :add_node
+
+      def inherit_data(track, name, data)
+        unless track[name].nil?
+          data[:facts] = track[name].facts.merge(data[:facts])
+          data[:vars] = track[name].vars.merge(data[:vars])
+          data[:features].concat(track[name].features)
+          inherit_data(track, track[name].name, data)
+        end
+        data
+      end
+      private :inherit_data
+
+      def create_target(target_name)
+        data = @groups.data_for(target_name) || {}
+        name_opt = {}
+        name_opt['name'] = data['name'] if data['name']
+
+        # If there is no name then this node was only referred to as a string.
+        uri = data['uri']
+        uri ||= target_name unless data['name']
```

Comment:
I think this needs to get set to host if URI isn't set and 'name' is defined, otherwise `target.uri` returns nil from the plan.